test/core/test_switching.py

```diff
@@ -0,0 +1,574 @@
+from fixate.core._switching import (
+    _generate_bit_sets,
+    VirtualMux,
+    _bit_generator,
+    PinSetState,
+    PinUpdate,
+    VirtualSwitch,
+    RelayMatrixMux,
+    PinValueAddressHandler,
+    generate_pin_group,
+    generate_relay_matrix_pin_list,
+    AddressHandler,
+    VirtualAddressMap,
+    MuxGroup,
+    JigDriver,
+)
+
+import pytest
+
+################################################################
+# helper to generate data
+
+
+def test_generate_bit_sets_empty():
+    assert list(_generate_bit_sets([])) == []
+
+
+def test_generate_bit_sets_one_bit():
+    assert list(_generate_bit_sets(["b0"])) == [set(), {"b0"}]
+
+
+def test_generate_bit_sets_multiple_bits():
+    expected = [
+        set(),
+        {"b0"},
+        {"b1"},
+        {"b1", "b0"},
+        {"b2"},
+        {"b2", "b0"},
+        {"b2", "b1"},
+        {"b2", "b1", "b0"},
+    ]
+    assert list(_generate_bit_sets(["b0", "b1", "b2"])) == expected
+
+
+def test_bit_generator():
+    """b1, b10, b100, b1000, ..."""
+    bit_gen = _bit_generator()
+
+    actual = [next(bit_gen) for _ in range(8)]
+    expected = [1, 2, 4, 8, 16, 32, 64, 128]
+    assert actual == expected
+
+
+def test_generate_pin_group():
+    assert list(generate_pin_group(3)) == (
+        "3K1 3K2 3K3 3K4 3K5 3K6 3K7 3K8 3K9 3K10 3K11 3K12 3K13 3K14 3K15 3K16".split()
+    )
+    assert list(generate_pin_group(1, prefix="RM")) == (
+        "RM1K1 RM1K2 RM1K3 RM1K4 RM1K5 RM1K6 RM1K7 RM1K8 RM1K9 RM1K10 RM1K11 RM1K12 RM1K13 RM1K14 RM1K15 RM1K16".split()
+    )
+    assert list(generate_pin_group(10, pin_count=8, prefix="U")) == (
+        "U10K1 U10K2 U10K3 U10K4 U10K5 U10K6 U10K7 U10K8".split()
+    )
```

Comment:
Each parameter is explicitly tested except for `sep`. Could add this for completeness.
```
assert list(generate_pin_group(10, pin_count=8, prefix="RM", sep="_")) == (
    "RM10_K1 RM10_K2 RM10_K3 RM10_K4 RM10_K5 RM10_K6 RM10_K7 RM10_K8".split()
)
```

Reply:
never mind, just found an instance below where sep is included.